Allow nightly builds of OSX binary

tbg opened the pull request:
Triggered via a master build with the `NIGHTLY` environment
variable. This ran in 10-15min, so no caching should be
necessary.

Still playing with this, but take a look if you like, especially
the Makefile.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4343)

<!-- Reviewable:end -->

mberhault commented on build/push-aws.sh:
```diff
@@ -14,30 +14,16 @@
-push_one_binary sql/sql.test
-push_one_binary acceptance/acceptance.test
-push_one_binary static-tests.tar.gz
+push_one_binary cockroach cockroach-${OSARCH}
```
you're passing two arguments to a 1-arg function.

Also, please hold before merging this, or it will break all the nightly stuff.

mberhault replied:
nevermind, I see what you did with the two args.

tschottdorf replied:
Yep. Don't worry, won't be merging this before updating all that rely on
this. For now, I'm just playing around and checking that it works.

On Thu, Feb 11, 2016 at 9:15 PM marc <notifications@github.com> wrote:

> In build/push-aws.sh
> <https://github.com/cockroachdb/cockroach/pull/4343#discussion_r52699008>:
>
> >
> > -push_one_binary cockroach
> > -push_one_binary sql/sql.test
> > -push_one_binary acceptance/acceptance.test
> > -push_one_binary static-tests.tar.gz
> > +push_one_binary cockroach cockroach-${OSARCH}
>
> nevermind, I see what you did with the two args.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4343/files#r52699008>.
>

mberhault commented on circle.yml:
```diff
@@ -39,3 +39,15 @@ deployment:
+            docker pull karalabe/xgo-develop
+            xgo --targets=darwin-10.9/amd64 --go=1.6rc2 --ldflags="$(make ldflags)" .
+            source build/build-common.sh
+            push_one_binary cockroach-darwin-10.9-amd64
```
you need `aws configure ...` in here somewhere

tamird commented:
:metal: 

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-:-KAINy27TK_IL-LJf76C:1147565816)*
<!-- Sent from Reviewable.io -->

bdarnell commented on Makefile:
```diff
@@ -63,6 +63,14 @@ build: GOFLAGS += -i -o cockroach
 build: BUILDMODE = build
 build: install
 
+# TODO(tschottdorf): how to avoid this repetition with `install`?
```
I think you could have the three `-X` flags generated in `build/ldflags.sh` and then use `install: LDFLAGS += $(shell build/ldflags.sh)`

bdarnell commented on circle.yml:
```diff
@@ -39,3 +39,23 @@ deployment:
+            export repo="${GOPATH%%:*}/src/github.com/cockroachdb/cockroach"
+            cd "${repo}/.."
+            rm -r "${repo}"
+            mv ~/cockroach "${repo}"
```
If the repo isn't in the right place, then we don't have our dependencies either. Is `xgo` running `go get` for us (in which case we're not getting our pinned dependency versions)

Oh - `make ldflags` will bootstrap. Sneaky.

tschottdorf replied:
Everything's already bootstrapped at this point: the GOPATH is mounted into the container, so the tests have already taken care of it.

bdarnell commented on circle.yml:
```diff
@@ -39,3 +39,23 @@ deployment:
+      - |
+          if [ -n "${BUILD_OSX-}" ]; then
+            set -eux
+            # There must certainly be a better way to move ~/cockroach
```
Why do we have to move it at all? Why can't we reuse the source from the tests we've already run?

tamird replied:
Actually, this is a valid point. I think we need to docker-in-docker to
make that work, but once we're inside the builder container, all the
directories are mounted correctly and no move is necessary.

On Sat, Feb 13, 2016 at 12:18 PM, Ben Darnell <notifications@github.com>
wrote:

> In circle.yml
> <https://github.com/cockroachdb/cockroach/pull/4343#discussion_r52830582>:
>
> > @@ -39,3 +39,23 @@ deployment:
> >            -l "${CIRCLE_ARTIFACTS}"/acceptance_deploy 2>&1 >
> >            "${CIRCLE_ARTIFACTS}/acceptance_deploy.log"
> >        - build/push-aws.sh
> > +  osx:
> > +    branch: master
> > +    commands:
> > +      - |
> > +          if [ -n "${BUILD_OSX-}" ]; then
> > +            set -eux
> > +            # There must certainly be a better way to move ~/cockroach
>
> Why do we have to move it at all? Why can't we reuse the source from the
> tests we've already run?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/4343/files#r52830582>.
>

tschottdorf replied:
We move it only so that it's in the correct location in GOPATH. We do reuse the source from tests we've already run - those mount `~/cockroach` where I'm moving it to, and the rest of GOPATH is mounted in separately, so it already gets a bootstrapped setup.

I had briefly looked at running it inside of the builder, but then we need the builder to be docker-in-docker as @tamird said, plus we need to run the builder in privileged mode (which looked like CircleCI potentially allowed it after explanation of the use case). But that seems too complicated for its own sake.

Could we just move `~/cockroach` to that location in the setup phase?

tamird replied:
@tschottdorf is it really that complicated? wouldn't it just boil down to `DOCKERFLAGS=--priviledged build/builder.sh xgo ...`?

tschottdorf replied:
No, the builder doesn't have Docker, so we need to integrate https://github.com/jpetazzo/dind/blob/master/Dockerfile (which works), but then that has another magic wrapper script https://raw.githubusercontent.com/jpetazzo/dind/master/wrapdocker; I am just testing this locally and it looks like it would probably work (if the `--privileged` flag works on CI), but there are various warnings.

I still don't think we should go down that route - it seems prone to surprising and intransparent breakage. So, what speaks against just moving `~/cockroach` in its "correct" location for all the tests?

bdarnell replied:
I think moving the directory is fine; I'm just surprised we would do it here instead of at `checkout: post:` above.

tschottdorf replied:
That's what I'm suggesting above - I'll see whether that works out.

bdarnell commented on circle.yml:
```diff
@@ -39,3 +39,23 @@ deployment:
+            mv ~/cockroach "${repo}"
+            cd "${repo}"
+            go get github.com/karalabe/xgo
+            xgo --targets=darwin-10.9/amd64 --go=1.6rc2 --ldflags="$(build/builder.sh make ldflags)" .
```
Add a comment about why `darwin-10.9`, even if it's just "10.9 is the most recent version available at this time".

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[circle.yml, line 10 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASU57fsFFunfLGZvfJ:-KASU57gqN3SA0dBSe_r:491504406) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L10)):*
Why the `&&` for the first 2 commands, but a separate command for the `ln`? Curious if there was some rationale there. 

---

*[circle.yml, line 50 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASTPB3m6UiFA0mQscD:-KASTPB3m6UiFA0mQscE:-585166196) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L50)):*
It's nice to have most of the work performed by `circle.yml` done outside of `circle.yml` itself so that the same scripts can be invoked by hand for development/debugging. Will the commands below work on our Mac's? If yes, then lets move this into a `build/build-osx-binaries.sh` script. If no, add a comment here explaining that this is linux specific.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[circle.yml, line 10 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASU57fsFFunfLGZvfJ:-KATceemw0IfkJaySSjp:906378889) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L10)):*
No rationale. Changed.

---

*[circle.yml, line 50 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASTPB3m6UiFA0mQscD:-KATbTMv_Rybw_CzqoOM:-2023528562) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L50)):*
Done. It works on our macs, though that's not that interesting. The more interesting part is that it (should) work on Linux.

---

*[Makefile, line 66 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAIEet5TvtD3STJKIWK-r3-66:-KATbJ6zjS-QVdqV55zd:55334140) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7bd208dbd700520c64eb003504b29b0966375b8/Makefile#L66)):*
works like a charm.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@mberhault I changed the code to leave the linux binary names untouched for this PR. We should homogenize, though.

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-:-KATe1JL1litsKVi3DBo:-2008325815)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r3, 3 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[build/ldflags.sh, line 8 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KATectpyNwS_TdCXi8S:-KATectpyNwS_TdCXi8T:2132726933) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3e634a68ad947bea0bb7dd7d6575402c21e437/build/ldflags.sh#L8)):*
trailing newline

---

*[circle.yml, line 10 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASU57fsFFunfLGZvfJ:-KATeAzGDqEPf1BHk0TM:-817565813) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L10)):*
could you add a comment here?

---

*[circle.yml, line 51 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KATekndKRxErMPcZuWx:-KATekndKRxErMPcZuWy:831340019) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3e634a68ad947bea0bb7dd7d6575402c21e437/circle.yml#L51)):*
where's this file?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 3 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[circle.yml, line 51 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KATekndKRxErMPcZuWx:-KAVebXT3xlkcj0PYla1:530954456) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3e634a68ad947bea0bb7dd7d6575402c21e437/circle.yml#L51)):*
it's there now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[build/build-osx.sh, line 1 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAVez6C2JBThaD0KNRq:-KAVez6C2JBThaD0KNRr:1312327733) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a78cadc55c32bfcda70a9c638909b33fa036c0a/build/build-osx.sh#L1)):*
`/usr/bin/env sh` or @knz might be grumpy

---

*[build/build-osx.sh, line 7 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAVewXidMLYm5kq90-X:-KAVewXidMLYm5kq90-Y:2132726933) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a78cadc55c32bfcda70a9c638909b33fa036c0a/build/build-osx.sh#L7)):*
trailing newline

---

*[build/ldflags.sh, line 1 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAVfDA8H-k6x7bvVB1N:-KAVfDA8H-k6x7bvVB1O:-609584483) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a78cadc55c32bfcda70a9c638909b33fa036c0a/build/ldflags.sh#L1)):*
this one too

---

*[circle.yml, line 10 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KASU57fsFFunfLGZvfJ:-KAVerOAlDJU3p2wofOn:-813413588) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f8f5dd19217bc7e2a2cde0d59a919c0f14fec4/circle.yml#L10)):*
Ah I meant one explaining why the move is needed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[build/build-common.sh, line 11 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY3ylge-X9wg6omjz4:-KAY3ylge-X9wg6omjz5:-1019282229) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/build-common.sh#L11)):*
random trailing newlines

---

*[build/build-osx.sh, line 3 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY4D2e__OZJZ1RtrWk:-KAY4D2fH4bJ9PUVlPm7:1081318631) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/build-osx.sh#L3)):*
nit: add a newline after this for consistency

---

*[build/push-aws.sh, line 17 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY4Hus2_HbflkyfUaa:-KAY4Hus2_HbflkyfUab:1509817863) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/push-aws.sh#L17)):*
make this order consistent with ldflags.sh

---

*[build/push-one-binary.sh, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY4ORL7aAcEve_4jlV:-KAY4ORL7aAcEve_4jlW:-1026889540) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/push-one-binary.sh#L1)):*
does this need bash or can it be sh like the others?

---

*[build/push-one-binary.sh, line 2 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY4NIthelynjYNbQKn:-KAY4NIthelynjYNbQKo:360335598) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/push-one-binary.sh#L2)):*
newline before this

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-:-KAY4ZmTslheugbSy6kL:980077423)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[build/push-one-binary.sh, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY4ORL7aAcEve_4jlV:-KAY5WSueEJ5HDYKxPQ8:1759478902) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0805aa782900173b1da8ab44d5e89c3c977dfd1/build/push-one-binary.sh#L1)):*
Added comment:

> This file uses `bash` and not `sh` due to the `time` builtin (the external
 `time` is not available on CircleCI).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r8.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[build/push-aws.sh, line 20 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY74-BJ02PnRQbZgF3:-KAY74-CWtyRlXIlR_HT:-1572377461) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e47864339bd2b4f017b1b1536ea4bae2249e7e/build/push-aws.sh#L20)):*
doesn't this have to precede the `git` invocation above?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 5 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tbg committed:
Allow nightly builds of OSX binary

Triggered via a master build with the `NIGHTLY` environment variable. The full
build should run in 10-15min, so no caching should be necessary.

tschottdorf commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[build/push-aws.sh, line 20 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343#-KAY74-BJ02PnRQbZgF3:-KAY7Ql-W8r3abrIgaRP:329316273) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6e47864339bd2b4f017b1b1536ea4bae2249e7e/build/push-aws.sh#L20)):*
yep. Brainless moving around.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4343)*
<!-- Sent from Reviewable.io -->
